src/SourceLocator/Ast/FindReflectionsInTree.php

```diff
@@ -41,94 +43,64 @@ public function __invoke(
-            private $astConversionStrategy;
-
-            /** @var Namespace_|null */
-            private $currentNamespace;
-
-            public function __construct(
-                Reflector $reflector,
-                IdentifierType $identifierType,
-                LocatedSource $locatedSource,
-                AstConversionStrategy $astConversionStrategy
-            ) {
-                $this->reflector             = $reflector;
-                $this->identifierType        = $identifierType;
-                $this->locatedSource         = $locatedSource;
-                $this->astConversionStrategy = $astConversionStrategy;
-            }
-
-            public function enterNode(Node $node) : void
-            {
-                if ($node instanceof Namespace_) {
-                    $this->currentNamespace = $node;
-
-                    return;
-                }
 
-                if ($node instanceof Node\Stmt\ClassLike) {
-                    $classNamespace = $node->name === null ? null : $this->currentNamespace;
-                    $reflection     = $this->astConversionStrategy->__invoke($this->reflector, $node, $this->locatedSource, $classNamespace);
-
-                    if ($this->identifierType->isMatchingReflector($reflection)) {
-                        $this->reflections[] = $reflection;
+        $nodeTraverser = new NodeTraverser();
+        $nodeTraverser->addVisitor(new NameResolver());
+        $ast = $nodeTraverser->traverse($ast);
+
+        static $collectNamespaceOrphanSymbols;
+        static $collectNamespacedSymbols;
+
+        if($collectNamespacedSymbols === null){
+
+            $collectNamespaceOrphanSymbols = [
+                // do not enter Namespace_ but do collect it
+                new CollectByTypeInstructions(Namespace_::CLASS, true, []),
+
+                // collect ClassLike and keep searching in its stmts
+                new CollectByTypeInstructions(ClassLike::CLASS, true, ['stmts']),
+
+                // collect Function_ and keep searching in its stmts
+                new CollectByTypeInstructions(Function_::CLASS, true, ['stmts']),
+
+                // @TODO can add more instructions to blacklist bits of the AST where not to
+                // look, for example constant expressions, function arguments and parameters,
+                // etc.
+            ];
+
+            $collectNamespacedSymbols = [
+                new CollectByTypeInstructions(ClassLike::CLASS, true, ['stmts']),
+                new CollectByTypeInstructions(Function_::CLASS, true, ['stmts']),
+
+                // @TODO same as above
+            ];
+        }
+
+        $collectByType = new CollectByType();
+        $rootNodes = $collectByType->collect($collectNamespaceOrphanSymbols, [$ast]);
+        /** @var Namespace_[]|ClassLike[]|Function_ $rootNodes */
+
+        $reflections = [];
+        foreach ($rootNodes as $node) {
+            if ($node instanceof Namespace_) {
```

Comment:
This block has the same problem of the above, but with:

```php
What happens with a conditional declaration such as https://3v4l.org/QjtHs ?

```php
<?php

namespace A;

if (true) {
    function foo() {}
} else {
    function bar() {}
}

bar();
```

Reply:
it's supported. the traversal is complete like in the nodevisitor solution